### PR TITLE
fix: add keepSessionStateOnFailover property to allow preserving session state after failover

### PR DESCRIPTION
**Pull request by crystall-bitquill**
### Summary

Add keepAutoCommitValueOnFailover property to prevent auto commit values from being permanently reset

### Description

This PR adds a property, `keepSessionStateOnFailover`, to maintain the session state after a failover. This includes the autocommit, sessionMaxRows and transaction isolation values. Ordinarily, these values are reset after failover and the user will need to reconfigure the connection. When `keepSessionStateOnFailover` is set to true, the user will not have to manually reset the listed values.

Addresses #425

### CODE REVIEW

**Review by jebeaudet — COMMENTED** [review]
Thanks for the quick feedback!

**Review by karenc-bq — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -289,20 +289,21 @@ To the application, the logical connection will not appear to have changed, but
...
+| `keepAutoCommitValueOnFailover`       | Boolean |                                                    No                                                     | This parameter will allow connections to retain the auto commit value after failover. When `keepAutoCommitValueOnFailover` is set to false, connections will need to be reconfigured and the autoCommit value will need to be reset as seen in the examples [here](#08007---Transaction-Resolution-Unknown). When set to true, re-setting the autoCommit value is not necessary. This parameter is only necessary when autoCommit is set to false and the connection cannot be reconfigured by the user.                                                                                                                                                                                                                        | `false`                                                                                                                                                                                        |
```
[the remark this answers is not in the document]
Comment (crystall-bitquill) [reply README.md]: Thanks for your feedback, I've updated the README and have made changes to restore other session parameters.

File: src/test/java/testsuite/integration/container/AuroraMysqlFailoverIntegrationTest.java [hunk 2]
```diff
@@ -434,13 +435,56 @@ public void test_failoverTimeoutMs() throws SQLException, IOException {
...
-      
+
       assertEquals(MysqlErrorNumbers.SQL_STATE_UNABLE_TO_CONNECT_TO_DATASOURCE, e.getSQLState());
       long duration = invokeEndTimeMs - invokeStartTimeMs;
       assertTrue(duration < 15000); // Add in 5 seconds to account for time to detect the failure
     }
   }
 
+  @Test
+  public void test_keepAutoCommitParameter()
```
Comment (sergiyvamz) [inline src/test/java/testsuite/integration/container/AuroraMysqlFailoverIntegrationTest.java:446]: I'd split this test into two simple tests. One checks behaviour without keepSessionStateOnFailover setting, one checks logic with keepSessionStateOnFailover.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java [hunk 3]
```diff
@@ -248,6 +250,8 @@ public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLExcepti
     if (this.enableFailoverSetting) {
       initProxy();
     }
+
+    saveSessionState();
```
Comment (sergiyvamz) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java:254]: Session state may change after initial connection is open. I think all state transfer logic should be in transferSessionState().

Comment (sergiyvamz) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java:254]: If we need to save transaction isolation we need to catch "Connection.setTransactionIsolation" in execute()

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java [hunk 4]
```diff
@@ -586,9 +598,13 @@ protected void failoverWriter() throws SQLException {
     metricsContainer.registerFailoverConnects(true);
 
     // successfully re-connected to the same writer node
-    updateCurrentConnection(
-        failoverResult.getNewConnection(),
-        WRITER_CONNECTION_INDEX);
+    if (!keepSessionStateOnFailover) {
```
Comment (sergiyvamz) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java:601]: I'm not sure we need this code.

**Pull request merged**